src/NzbDrone.Core/Download/Clients/Aria2/Aria2.cs

```diff
@@ -0,0 +1,223 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using CookComputing.XmlRpc;
+using FluentValidation.Results;
+using NLog;
+using NzbDrone.Common.Disk;
+using NzbDrone.Common.Extensions;
+using NzbDrone.Common.Http;
+using NzbDrone.Core.Configuration;
+using NzbDrone.Core.MediaFiles.TorrentInfo;
+using NzbDrone.Core.Parser.Model;
+using NzbDrone.Core.RemotePathMappings;
+using NzbDrone.Core.Validation;
+
+namespace NzbDrone.Core.Download.Clients.Aria2
+{
+    public class Aria2 : TorrentClientBase<Aria2Settings>
+    {
+        private readonly IAria2Proxy _proxy;
+
+        public override string Name => "Aria2";
+
+        public Aria2(IAria2Proxy proxy,
+                        ITorrentFileInfoReader torrentFileInfoReader,
+                        IHttpClient httpClient,
+                        IConfigService configService,
+                        IDiskProvider diskProvider,
+                        IRemotePathMappingService remotePathMappingService,
+                        Logger logger)
+            : base(torrentFileInfoReader, httpClient, configService, diskProvider, remotePathMappingService, logger)
+        {
+            _proxy = proxy; ;
+        }
+
+        protected override string AddFromMagnetLink(RemoteEpisode remoteEpisode, string hash, string magnetLink)
+        {
+            var gid = _proxy.AddMagnet(Settings, magnetLink);
+
+            var tries = 10;
+            var retryDelay = 500;
+
+            // Wait a bit for the magnet to be resolved.
+            if (!WaitForTorrent(gid, hash, tries, retryDelay))
+            {
+                _logger.Warn($"Aria2 could not add magnent within {tries * retryDelay / 1000} seconds, download may remain stuck: {magnetLink}.");
+                return hash;
+            }
+
+            _logger.Debug($"Äria2 AddFromMagnetLink '{hash}' -> '{gid}'");
+
+            return hash;
+        }
+
+        protected override string AddFromTorrentFile(RemoteEpisode remoteEpisode, string hash, string filename, byte[] fileContent)
+        {
+            var gid = _proxy.AddTorrent(Settings, fileContent);
+
+            var tries = 10;
+            var retryDelay = 500;
+
+            // Wait a bit for the magnet to be resolved.
+            if (!WaitForTorrent(gid, hash, tries, retryDelay))
+            {
+                _logger.Warn($"Aria2 could not add torrent within {tries * retryDelay / 1000} seconds, download may remain stuck: {filename}.");
+                return hash;
+            }
+
+            return hash;
+        }
+
+        public override IEnumerable<DownloadClientItem> GetItems()
+        {
+            var statuses = _proxy.GetStatuses(Settings);
```

Comment:
I think this should be renamed to torrents/GetTorrents, it'll be more obvious what it is at a glance, the proxy can handle the translation to Aria2 logic.

Particularly `foreach(var status in statuses)` is confusing, why am I looping through statuses and then getting a separate status from the status.